File: algorithms-week5/money_change_again/python/change_dp.py
# Uses python3
import sys

def get_change(m):
    NumCoins = 0
    MinNumCoins = []
    MinNumCoins.append(0)
    Coins = [1, 3, 4]

    for i in range(1, m+1):
        MinNumCoins.append(i + 1)

        for coin in list(Coins):
            if i >= coin:
                NumCoins = MinNumCoins[i-coin] + 1

            if NumCoins < MinNumCoins[i]:
                MinNumCoins[i] = NumCoins

    return MinNumCoins[m]

if __name__ == '__main__':
    m = int(sys.stdin.read())
    print(get_change(m))

File: algorithms-week5/money_change_again/cpp/change_dp.cpp
//
// Created by Sam on 8/29/2018.
//

#include <iostream>
#include <vector>

using namespace std;

int get_change(int m) {
  int NumCoins = 0;
  int MinNumCoins[m+1];
  MinNumCoins[0] = 0;

  vector<int> denominator = {1, 3, 4};

  for (int i = 1; i <= m; i++) {
    MinNumCoins[i] = i + 1;

    for (const int& coin : denominator) {
      if (i >= coin) {
        NumCoins = MinNumCoins[i - coin] + 1;
        if (NumCoins < MinNumCoins[i]) {
          MinNumCoins[i] = NumCoins;
        }
      }
    }
  }

  return MinNumCoins[m];
}

int main() {
  int m;
  std::cin >> m;
  std::cout << get_change(m) << '\n';
}
